### PR TITLE
Fixes #28405 - Missing permissions to Table Preferences

### PR DESCRIPTION
**Pull request by domitea**
Fixing bug where user was unable to see own table preferences via API.

### CODE REVIEW

File: app/registries/foreman/access_permissions.rb [hunk 1]
```diff
@@ -757,4 +757,9 @@
     map.permission :edit_settings, { :settings => [:update],
                                      :'api/v2/settings' => [:update] }
   end
+
+  permission_set.security_block :table_preferences do |map|
+    map.permission :view_table_preferences, { 'api/v2/table_preferences': [:index, :show] }
+    map.permission :edit_table_preferences, { 'api/v2/table_preferences': [:create, :update, :destroy] }
```
Comment (xprazak2) [inline app/registries/foreman/access_permissions.rb]: Is there a specific reason to group :create, :update and :destroy? We do not do that for any other resources but have a separate permission for each, so I am just wondering whether it makes sense.

**Comment by xprazak2** [comment]
Works as expected, I left one comment and there are related [test failures](https://ci.theforeman.org/job/test_develop_pr_core/2894/database=postgresql,ruby=2.5,slave=fast/testReport/junit/(root)/SeedHelperTest/test_0006_Does_not_fail_on_modified_default_role_with_filter/)

**Review by tbrisker — CHANGES_REQUESTED** [review]
There is no permission for these on purpose - they should be always accessible to the user. You can check the discussion at https://github.com/theforeman/foreman/pull/5262#discussion_r174460333 and around it. I would suggest instead finding where this permission check is failing and fix it there.

**Comment by tbrisker** [comment]
hint: https://github.com/theforeman/foreman/blob/develop/app/registries/foreman/access_permissions.rb#L5-L12 might give an idea to how to fix it. Might also be good to add a test to make sure the endpoint is accessible to non-admin user.

**Comment by domitea** [comment]
I'm little confused, I've discussing this issue with @ares and he recommends to fix it in `editing_self?` method in `User` model because user have to be able to make CRUD operations at own settings (in that case) so @tbrisker what is better to you? Use your suggested solution or edit `editing_self?` method?

**Comment by domitea** [comment]
PR is not ready for review. It's WIP

File: app/models/user.rb [hunk 2]
```diff
@@ -422,6 +422,52 @@ def can_change_admin_flag?
...
+        action: /show|destroy|index|create/,
+        cond: ->(options) { options[:user_id].to_i == self.id },
+      },
+    ]
+
+    complete_bool = false
+    controller_actions.each do |action_hash|
+      return true if complete_bool
+
+      controller_bool, action_bool = [:controller, :action].each do |symbol|
```
Comment (ezr-ondrej) [inline app/models/user.rb]:
```suggestion
      controller_matches, action_matches = [:controller, :action].map do |resource|
```

File: app/models/user.rb [hunk 3]
```diff
@@ -422,6 +422,52 @@ def can_change_admin_flag?
...
+      {
+        controller: 'api/v2/personal_access_tokens',
+        action: /show|destroy|index|create/,
+        cond: ->(options) { options[:user_id].to_i == self.id },
+      },
+    ]
+
+    complete_bool = false
+    controller_actions.each do |action_hash|
+      return true if complete_bool
```
Comment (ezr-ondrej) [inline app/models/user.rb]:
```suggestion
```

File: app/models/user.rb [hunk 4]
```diff
@@ -422,6 +422,52 @@ def can_change_admin_flag?
...
+      controller_bool, action_bool = [:controller, :action].each do |symbol|
+        if (value = action_hash[symbol]).is_a?(Regexp)
+          value =~ options[symbol]
+        else
+          value.to_s == options[symbol]
+        end
+      end
+      cond_bool = action_hash[:cond].call(options)
+
+      complete_bool = controller_bool && action_bool && cond_bool
```
Comment (ezr-ondrej) [inline app/models/user.rb]:
```suggestion
      return true if controller_matches && action_matches && cond_matches
```

File: app/models/user.rb [hunk 5]
```diff
@@ -422,6 +422,52 @@ def can_change_admin_flag?
...
+      return true if complete_bool
+
+      controller_bool, action_bool = [:controller, :action].each do |symbol|
+        if (value = action_hash[symbol]).is_a?(Regexp)
+          value =~ options[symbol]
+        else
+          value.to_s == options[symbol]
+        end
+      end
+      cond_bool = action_hash[:cond].call(options)
```
Comment (ezr-ondrej) [inline app/models/user.rb]:
```suggestion
      cond_matches = action_hash[:cond].call(options)
```

**Review by ezr-ondrej — COMMENTED** [review]
I believe it's the best we can achieve with current implementation, although I would rather see our permissions to allow conditions themselves.

File: app/models/user.rb [hunk 6]
```diff
@@ -422,6 +422,52 @@ def can_change_admin_flag?
...
+
+      complete_bool = controller_bool && action_bool && cond_bool
+    end
+
+    return false
```
Comment (tbrisker) [inline app/models/user.rb]: All code after this point should be removed i guess?

**Review by tbrisker — COMMENTED** [review]
I'd like to see a test showing this actually fixes the issue. 
I also still believe it would be simpler to add this endpoint to the public security block, since the controller has a check to make sure a user can only see or edit their own preferences - https://github.com/theforeman/foreman/blob/develop/app/controllers/api/v2/table_preferences_controller.rb#L6.

**Comment by domitea** [comment]
Thanks @tbrisker for a good option how to fix it. I had to refactor `editing_self?` method and I was unable to use that method due to `user_aware` concern. So I added new public permissions and the test.

Is everything good now? @xprazak2 @tbrisker @ezr-ondrej

File: app/models/user.rb [hunk 7]
```diff
@@ -577,6 +577,36 @@ def authenticate_by_personal_access_token(token)
 
   private
 
+  def allowed_controller_actions
```
Comment (ezr-ondrej) [inline app/models/user.rb]: Definitely not a good name. It sounds pretty general, but it really is not.
Otherwise I do like the refactoring, but it's now unrelated to the issue, so it should be separate PR

File: app/registries/foreman/access_permissions.rb [hunk 8]
```diff
@@ -9,6 +9,10 @@
       :notification_recipients => [:index, :update, :destroy, :update_group_as_read, :destroy_group ] }, :public => true
     map.permission :api_status, { :"api/v2/home" => [:status]}, :public => true
     map.permission :about_index, { :about => [:index] }, :public => true
+    map.permission :view_table_preferences, { 'api/v2/table_preferences': [:index, :show] }, public: true
```
Comment (ezr-ondrej) [inline app/registries/foreman/access_permissions.rb]: Do we need so many permissions for such a simple resource? :(

**Review by ezr-ondrej — COMMENTED** [review]
I'm personally in favor of adding this to `editing_self?` or use one general public permission for whole resource.

File: app/registries/foreman/access_permissions.rb (hunk 8 above)
Comment (ares) [reply app/registries/foreman/access_permissions.rb]: I'd also be pro refactoring `editing_self?` if possible instead of adding a new permission, users should be able to edit their preferences only and if e.g. admin needs to edit preferences of other users, they can impersonate their account

File: app/registries/foreman/access_permissions.rb [hunk 9]
```diff
@@ -9,6 +9,10 @@
       :notification_recipients => [:index, :update, :destroy, :update_group_as_read, :destroy_group ] }, :public => true
```
Comment (tbrisker) [inline app/registries/foreman/access_permissions.rb]:
```suggestion
       :notification_recipients => [:index, :update, :destroy, :update_group_as_read, :destroy_group ],
       :"api/v2/table_preferences" => [:show, :create, :edit, :delete, :index]}, :public => true
```

**Review by tbrisker — CHANGES_REQUESTED** [review]
I guess my suggestion wasn't clear, I don't think we need any new permissions or changes to the `editing_self?` method, just the change to make the table preferences endpoints public. 
I added a suggested change that from my testing seems to be enough to get the endpoint working for non-admin users.

**Comment by domitea** [comment]
Requested changes has taken into code. Is it ok now? @tbrisker @ezr-ondrej @xprazak2

File: db/seeds.d/020-permissions_list.rb [hunk 10]
```diff
@@ -186,6 +186,10 @@ def permissions
         ['User', 'edit_users'],
         ['User', 'destroy_users'],
         [nil, 'view_statuses'],
+        ['TablePreference', 'view_table_preferences'],
+        ['TablePreference', 'create_table_preferences'],
+        ['TablePreference', 'edit_table_preferences'],
+        ['TablePreference', 'destroy_table_preferences'],
```
Comment (ezr-ondrej) [inline db/seeds.d/020-permissions_list.rb]: I guess these are leftovers, we are not really using them now, right?

**Review by ezr-ondrej — COMMENTED** [review]
LGTM, just remove the permissions from seeding, while we decided not to use them and I think we are good to go :+1:

File: db/seeds.d/020-permissions_list.rb (hunk 10 above)
Comment (domitea) [reply db/seeds.d/020-permissions_list.rb]: Yes. Thanks!

**Comment by domitea** [comment]
@oezr is it ready for merge from your point of view?

**Review by tbrisker — DISMISSED** [review]
LGTM, can you please rebase on latest develop to get the tests to pass?

**Review by ezr-ondrej — APPROVED** [review]
Thanks @domitea, works as expected :+1:

**Comment by tbrisker** [comment]
1.24 - e301e5beb
